### PR TITLE
add actors events to itil objects

### PR DESCRIPTION
**Pull request by orthagh**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| old pr | #274 

Part of http://glpi.userecho.com/topics/128-behavior-plugin-integration/

### CODE REVIEW

File: inc/commonitilactor.class.php [hunk 1]
```diff
@@ -346,15 +346,31 @@ function post_addItem() {
...
+         }
+
+         // raise notification for this actor addition
+         $string_type = "";
+         switch ($this->input['type']) {
+            case self::REQUESTER:
+               $string_type = "requester";
+               break;
+            case self::OBSERVER:
+               $string_type = "obserrver";
```
Comment (tomolimo) [inline inc/commonitilactor.class.php]: need to change this to observer

File: inc/notificationtargetcommonitilobject.class.php [hunk 2]
```diff
@@ -55,6 +55,23 @@ function __construct($entity='', $event='', $object=null, $options=array()) {
       $this->options['sendprivate'] = true;
    }
 
+   /**
+    * Get events related to ItiObject
```
Comment (tomolimo) [inline inc/notificationtargetcommonitilobject.class.php]: typo: ItiObject -> ItilObject

File: inc/notificationtargetcommonitilobject.class.php [hunk 3]
```diff
@@ -55,6 +55,23 @@ function __construct($entity='', $event='', $object=null, $options=array()) {
...
+   **/
+   function getEvents() {
+
+      $events = array('requester_user'    => __('Addition of a user in requesters'),
+                      'requester_group'   => __('Addition of a group in requesters'),
+                      'observer_user'     => __('Addition of a user in observers'),
+                      'observer_group'    => __('Addition of a group in observers'),
+                      'assign_user'       => __('Assignation to a technician'),
```
Comment (tomolimo) [inline inc/notificationtargetcommonitilobject.class.php]: Harmonization of texts:
Addition of a user in technicians
Addition of a group in technicians
Addition of a supplier in technicians

File: inc/commonitilactor.class.php [hunk 4]
```diff
@@ -346,15 +346,31 @@ function post_addItem() {
...
+               break;
+            case self::ASSIGN:
+               $string_type = "assign";
+               break;
          }
+         // example for event: assign_group
+         $event = $string_type."_".strtolower($this::$itemtype_2);
+         NotificationEvent::raiseEvent($event, $item);
```
Comment (tomolimo) [inline inc/commonitilactor.class.php]: Here must be tested the `$CFG_GLPI["use_mailing"]` (to respect current code, even if it is also tested in the `raiseEvent` method) and at least the `$this>input["_disablenotif"]` flag to know if the `raiseEvent` should be done or not, see PR #1722

Comment (orthagh) [inline inc/commonitilactor.class.php]: ok for _disablenotif.
I don't get the point of duplicate $CFG_GLPI["use_mailing"].
If we have this kind of code in another place, i think we need to deduplicate

File: inc/notificationtargetcommonitilobject.class.php (hunk 3 above)
Comment (orthagh) [inline inc/notificationtargetcommonitilobject.class.php]: I'm not sure about this locales.
I don't see a supplier as technician.

**Comment by orthagh** [comment]
base notifications incoming into 9.2-empty.sql

**Comment by tomolimo** [comment]
In fact in the raiseEvent method, it starts with if($CFG_GLPI["use_mailing"]) {…}
So if we test this value before calling the raiseEvent, we have a double test…

**Comment by orthagh** [comment]
yes, that's my point, we need to clean theses double tests (and not adding new ones)

File: inc/notificationtargetcommonitilobject.class.php (hunk 3 above)
Comment (tomolimo) [inline inc/notificationtargetcommonitilobject.class.php]: then the best would be

User assigned as Technician
Group assigned as Technician
Supplier assigned as Technician

Comment (tomolimo) [inline inc/notificationtargetcommonitilobject.class.php]: Same for Requester and Observer
User assigned as Requester
Group assigned as Requester
...

**Comment by tomolimo** [comment]
The problem is that it should be tested in the raiseEvent either for the use_mailing or for the _disablenotif, but in the raiseEvent, we don't have any options  to pass the _disablenotif...

File: inc/notificationtargetcommonitilobject.class.php (hunk 3 above)
Comment (orthagh) [inline inc/notificationtargetcommonitilobject.class.php]: No, we can't talk about assignment and requester in the same sentence, i think.
My problem is we don't have a common name for attribution like requester and observer.

My thought upper in this conversation is i found strange to use technician and supplier in the same sentence. Attribution is better i think.

Maybe we can stay for the initial proposal for requester and observer and use this sentence for the rest:
"Addition of a _type in attribution"
or
"Addition of a _type in assignees"

**Comment by orthagh** [comment]
for the purpose of this pr, i think we can stay on use_mailing in raiseEvent and prefix call by _disablenotif.

In another one (#1722 ?), we could think about adding a $params in raiseEvent to pass some aditionnal options, no ?

**Comment by tomolimo** [comment]
Ok for both proposals

**Comment by orthagh** [comment]
Pr up to date

**Comment by tomolimo** [comment]
Ok, I'll do like you proposed in PR #1722

**Comment by orthagh** [comment]
Great !

**Comment by orthagh** [comment]
IMO, no.
An actor of ticket is never updated (glpi_tickets_users), any change on this table always results on:
- an optional suppression of the old actor
- and addition of the new actor

**Comment by trasher** [comment]
The `empty.sql` file has been changed, and there is nothing regarding the update.

**Comment by orthagh** [comment]
ah, sorry...

**Comment by orthagh** [comment]
so no also, by the past (and i think it's a good way), we didn't add any new notifications to existing instances to avoid unwanted mailing.

**Comment by trasher** [comment]
So users who upgrade will never have the new notifications? Is taht not possible to add them, but disabled?

**Pull request merged**